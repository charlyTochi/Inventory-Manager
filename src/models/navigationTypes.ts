import {NativeStackScreenProps} from '@react-navigation/native-stack';

export type ListItemStackParamList = {
  Splash: undefined;
  Login: undefined;
  ListItems: undefined;
  CreateItem: undefined;
  ViewItem: {id: string};
  EditItem: {id: string};
};


export type LoginScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'Login'
>;

export type ListItemsScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'ListItems'
>;
export type CreateItemScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'CreateItem'
>;
export type ViewItemScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'ViewItem'
>;
export type EditItemScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'EditItem'
>;
