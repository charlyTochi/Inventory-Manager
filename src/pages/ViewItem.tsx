import React, {useCallback, useEffect, useState} from 'react';
import {Alert, Button, Dimensions, StyleSheet, Text, View} from 'react-native';
import {IInventoryItem, Inventory} from '../services/inventory';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RouteProp, useNavigation} from '@react-navigation/native';
import {ListItemStackParamList} from '../models/navigationTypes';

type ViewItemScreenNavigationProp = NativeStackNavigationProp<
  ListItemStackParamList,
  'ViewItem'
>;
type ViewItemScreenRouteProp = RouteProp<ListItemStackParamList, 'ViewItem'>;

type Props = {
  navigation: ViewItemScreenNavigationProp;
  route: ViewItemScreenRouteProp;
};

const inventory = new Inventory();

const ViewItem: React.FC<Props> = ({navigation, route}) => {
  const [item, setItem] = useState<IInventoryItem | null>(null);
  const navigationHook = useNavigation();
  const fetchItem = async () => {
    setItem(await inventory.read(route.params.id));
  };
  useEffect(() => {
    const handleScreenFocus = () => {
      fetchItem();
    };

    // Subscribe to the focus event
    const unsubscribe = navigationHook.addListener('focus', handleScreenFocus);

    // Clean up the listener when the component is unmounted
    return () => {
      unsubscribe();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [navigationHook]);

  const handleDeleteItem = useCallback(() => {
    Alert.alert(
      'Delete Item',
      'Are you sure you want to delete this item?',
      [
        {
          text: 'Cancel',
          onPress: () => console.log('Cancel Pressed'),
          style: 'cancel',
        },
        {
          text: 'OK',
          onPress: async () => {
            if (item) {
              await inventory.delete(item.id);
              navigation.goBack();
            }
          },
        },
      ],
      {cancelable: false},
    );
  }, [item, navigation]);

  useEffect(() => {
    fetchItem();
    return () => {
      fetchItem();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (!item) {
    return <Text>Loading...</Text>;
  }

  return (
    <View style={styles.container}>
      <View style={styles.itemContainer}>
        <Text>Name</Text>
        <Text style={styles.body}>{item.name}</Text>
        <Text>Description</Text>
        <Text style={styles.body}>{item.description}</Text>
        <Text>Total Stock</Text>
        <Text style={styles.body}>{item.totalStock}</Text>
        <Text>Price</Text>
        <Text style={styles.price}>
          {inventory.formatCurrency(Number(item?.price?.toFixed(2)))}
        </Text>
      </View>
      <Button
        title="Edit Item"
        onPress={() => navigation.navigate('EditItem', {id: item.id})}
      />
      <Button title="Delete Item" color={'red'} onPress={handleDeleteItem} />
    </View>
  );
};

const styles = StyleSheet.create({
  itemContainer: {flex: 1, marginTop: 30},
  body: {
    fontSize: 20,
    marginBottom: 20,
    textTransform: 'capitalize',
  },
  price: {
    fontSize: 20,
    marginBottom: 10,
  },
  container: {
    marginHorizontal: 10,
    flexDirection: 'column',
    height: Dimensions.get('screen').height - 150,
    gap: 5,
  },
});
export default ViewItem;
