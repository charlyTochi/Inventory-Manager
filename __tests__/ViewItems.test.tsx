import {render} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';
import ViewItem from '../src/pages/ViewItem';
import {ListItemStackParamList} from '../src/models/navigationTypes';
import {RouteProp} from '@react-navigation/native';
import React from 'react';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));

describe('ViewItem screen', () => {
  const mockNavigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setOptions: jest.fn(),
    addListener: jest.fn(),
    removeListener: jest.fn(),
    reset: jest.fn(),
    pop: jest.fn(),
    popToTop: jest.fn(),
    dispatch: jest.fn(),
  };

  const mockRoute: RouteProp<ListItemStackParamList, 'ViewItem'> = {
    key: 'ViewItem',
    name: 'ViewItem',
    params: {id: '1'},
  };

  it('should match the snapshot', () => {
    const {toJSON} = render(
      <NavigationContainer>
        <ViewItem navigation={mockNavigation as any} route={mockRoute} />
      </NavigationContainer>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
