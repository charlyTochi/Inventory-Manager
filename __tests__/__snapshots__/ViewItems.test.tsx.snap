// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewItem screen should match the snapshot 1`] = `
<Text>
  Loading...
</Text>
`;
